vLLM language model wrapper

Scezaquer opened the pull request:
Hello,

If I am not mistaken, vLLM is not currently supported by concordia. According to [this file](https://github.com/google-deepmind/concordia/blob/main/concordia/language_model/README.md), I assume this is a desired addition. I see there is an other PR about this (#162) which seems dead as the contributor has not signed the CLA since last month, which is why I am opening this new PR instead.

This PR implements a language model wrapper for vLLM following the instructions in [concordia/language_model/README.md](https://github.com/google-deepmind/concordia/blob/main/concordia/language_model/README.md).

As this is my first contribution to the project, please let me know if I have done something incorrectly.

locross93 commented:
Thanks so much for this contribution! You're right, vLLM support is a great addition, and we really appreciate you taking the initiative to implement it. This looks good and is close to being ready. We just need a few changes:

1. In setup.py, vllm is currently listed in the main install_requires. This would be a breaking change, as it would force all Concordia users to install the very large vllm package, which has specific CUDA requirements. Could you please move vllm from `install_requires` to the `extras_require` dictionary? Something like this:

`extras_require={
    'vllm': ['vllm'],
    # ... other existing extras in 'dev'
},`

2. There's a subtle bug in the `sample_choice` implementation. If I’m understanding vLLM’s documentation correctly, by setting `SamplingParams(prompt_logprobs=20)`, you are only retrieving the log probabilities for the top 20 most-likely tokens at each position. If a token in the choice string is not in that top-20 list (e.g., it's the 21st most-likely), it won't be found, and its logprob won't be added to the total. This will cause the code to incorrectly score that response. Change to `prompt_logprobs=None`.

3. I noticed in `sample_text` that you're capping the output using `min(max_tokens, _DEFAULT_MAX_TOKENS)`, which sets a hard limit of 1000 tokens. The standard Concordia interface default in language_model is 5000, so this might be unexpectedly low for users who request, for example, 2000 tokens. You could remove the `min(...) ` wrapper and just pass the max_tokens variable from the function argument directly to `SamplingParams`.

Scezaquer commented:
I believe the problems you identified should now be fixed.

1. Moved vllm from `install_requires` to `extras_require`. I re-generate `requirements.txt` using the script in `/bin/update_requirements.sh`, which seems to make a lot of changes? It's unclear to me if that's fine or not. I also updated `/bin/update_requirements.sh` to include the vllm extra.

2. You are correct that 20 is the wrong value, however looking at the docs [here](https://docs.vllm.ai/en/v0.6.5/dev/sampling_params.html) under `logprobs` it is stated that the logprob of the sampled token will always be returned. After testing, I can confirm that this also applies to `prompt_logprobs`. This means that having `prompt_logprobs=20` was fine and the token in the choice string will always be in the list even if it is not in the top 20. However this also means we can actually set `prompt_logprobs=0` to only get the logprob of the tokens we are interested in. After testing I can confirm this works, so I implemented this change. Note that setting `prompt_logprobs=None` would in fact turn-off prompt_logprobs entirely and break the method.

3. Set correct value for max_token.

I also added two arguments to VLLMLanguageModel (enable_prefix_caching and max_lora_rank) that I believe may be useful (or at the very least that I plan on using).

Thanks again for the feedback and please let me know if there is anything else I should change before this gets merged

Scezaquer commented:
I have separated `VLLMLanguageModel` into `VLLMLanguageModel` and `VLLMLora`. This is in case someone wants to load multiple LoRAs for the same underlying model: it would allow them to only instantiate `VLLMLanguageModel` once, and pass it by reference to as many `VLLMLora` as needed. This means the model is only loaded in memory once, instead of having to load a separate instance for every LoRA, which would be exceedingly wasteful.

locross93 reviewed (commented):
Great thanks for fixing these!
